docs/porting/target/static_pinmap.md

```diff
@@ -0,0 +1,180 @@
+<h1 id="static-pinmap-port">Static Pinmap extension</h1>
+
+The **Static Pinmap extension** allows you to statically specify the peripheral configuration (pin, peripheral or function) in the HAL API function.
+
+In modern MCUs, you can often map peripherals to different pins, and each pin can have multiple functions. Mbed supports dynamic pin mapping, meaning you can reconfigure pins at run time for different drivers to use. That provides great flexibility, but it's not free. There's a nontrivial ROM cost to maintain the pin map tables and infrastructure to parse it. In some use cases, this flexibility is worth the cost. Often, pin configuration is frozen at the hardware design stage and doesn't require run time modification. Shifting this configuration to compile time allows free memory associated with the dynamic approach.
+
+HAL APIs using pins take these pins in their constructor and use those pins to look up which peripheral or function to use. The process of looking up the peripheral or function requires there to be a pin map table that maps pins to peripherals or functions. This pin map table takes up ROM. Static pinmap extension provides additional HAL API constructors, which take pin map as a parameter where the pin, peripheral or function is specified statically, and there is no need to use the pin map tables.
+
+Supported peripherals:
+
+ - `PWM`.
+ - `AnalogIn`.
+ - `AnalogOut`.
+ - `SPI`.
+ - `I2C`.
+ - `UART`.
+ - `QSPI`.
+ - `CAN`.
+ 
+## Assumptions
+
+1. Provide types that will hold static pin maps for peripherals (`PWM`, `AnalogIn`, `AnalogOut`, `SPI`, `I2C`, `UART`, `QSPI`, `CAN`).
+2. Provide `xxx_init_direct(xxx_t *obj, static_pinmap_t *)` functions to HAL API (these functions won't use pin map tables).
+3. Provide additional constructors in the drivers layer, which will use the `xxx_init_direct(xxx_t *obj, static_pinmap_t*)` HAL functions.
+4. Provide default weak implementations of `xxx_init_direct(static_pinmap_t *)` functions. These functions call standard `xxx_init(xxx_t *obj, PinName, ...)` function (backward compatibility for targets that don't support the static pin map mechanism).
+5. Provide `constexpr` utility functions to lookup for pin mapping in compile time (requires C++14).
+6. Provide `constexpr` pin map tables in the header file.
+7. Provide macros for the pin map tables.
+8. Provide `STATIC_PINMAP_READY` macro in `PinNames.h`.
+
+## Implementing static pin map extension
+
+Most of the above points are already implemented. To make static pin map available on your platform, please perform the following steps: 
+
+- Provide implementation of `xxx_init_direct(xxx_t *obj, static_pinmap_t *)` function (which does not use pin map tables).
```

Comment:
Do these bulleted steps need to happen in order?

Reply:
Nope, but all together :-)